src/library/export/engineprimeexportjob.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <QAtomicInteger>
+#include <QList>
+#include <QMutex>
+#include <QQueue>
+#include <QSet>
+#include <QThread>
+#include <QWaitCondition>
+#include <memory>
+
+#include "library/export/engineprimeexportrequest.h"
+#include "library/trackcollectionmanager.h"
+#include "library/trackset/crate/crate.h"
+#include "library/trackset/crate/crateid.h"
+#include "track/track.h"
+#include "track/trackid.h"
+#include "track/trackref.h"
+
+namespace mixxx {
+
+/// The Engine Prime export job performs the work of exporting the Mixxx
+/// library to an external Engine Prime (also known as "Engine Library")
+/// database, using the libdjinterop library, in accordance with the export
+/// request with which it is constructed.
+class EnginePrimeExportJob : public QThread {
```

Comment:
Please think about separating the worker thread from the class that runs on the UI event loop. The worker thread should not block the UI thread.

Reply:
My understanding was that all the interaction/loading with `TrackCollectionManager` had to take place on the main thread - we even had prior discussion on this PR about this topic.  Is this still the case?

For this reason, the ctor calls `moveToThread(m_pTrackCollectionManager->thread());`, so that the private slots on `EnginePrimeExportJob` use the main thread.

If tracks/crates/etc. have to be loaded on the main thread, how do you propose loading can be separated?

Reply:
Just ensure that the UI thread cannot be blocked by the worker thread. In your design the UI thread needs to wait for the worker thread to release the mutex causing a dependency inversion (assuming the worker thread is scheduled with lower priority).

Either use a worker thread with a message loop and signal/slot connections or pass data in a lock-free manner. The worker thread should wait for the UI thread, but not the other way round.

Reply:
Ok - it is correct to say that the UI thread can become blocked waiting on the worker thread.  The worker thread does something like this:

```cpp
void workerThreadMethod() {
  ...
  QMutexLocker lock{&m_mutex};
  QMetaObject::invokeMethod(this, "runInMainThread", ...);
  m_waitCondition.wait(&m_mutex);
  ...
}
```

And the UI thread slot does this:
```cpp
void runInMainThread(...) {
  // Do work that must happen in Main/UI thread...

  QMutexLocker lock{&m_mutex};
  m_waitCondition.wakeAll();
}
```

However, this was deliberate: locking the mutex in `runInMainThread()` avoids the "lost wakeup" problem, i.e. it blocks until the wait condition is definitely being waited upon.  If it didn't, the method could finish before the worker thread for to the wait condition, and then it would wait forever.

I could try having the UI thread write to an atomic integer to signal that its work is done _in addition to_ calling `QWaitCondition::wakeAll()` but _instead of_ using the mutex, and then the `QWaitCondition::wait()` method in the worker thread can use a short time-out via the `QDeadlineTimer` overload, so that lost wake-up can be avoided by periodically checking the atomic integer.  I'll cross my fingers that lost wake-up probably wouldn't happen that often, so the time-out route won't penalise the export too much.

A _totally_ lock-free route would probably involve the worker thread polling a `QAtomicInteger` on a delay loop (with no wait condition), but it feels a little messy to be using it in this way.

I'll give it a try with `QWaitCondition` and `QAtomicInteger` together...

Reply:
The multi-threaded analysis uses a single slot lock-free queue for handing over a task to a worker thread and the worker threads emit a signal for reporting progress in the opposite direction.

Reply:
I pushed a far simpler approach: the worker thread now uses `Qt::BlockingQueuedConnection` to treat the main-thread slot invocation like a normal blocking method call.  This is safe as the slots are guaranteed to be on a different thread from the worker.

This keeps the `run()` method fairly intuitive to read from start to finish.  There are no longer any mutexes or wait conditions in the export job class to complicate the readability.

Reply:
Sounds reasonable :+1: